File: src/main/java/lzhang/leetcode/Three.java
package lzhang.leetcode;

import java.util.HashMap;

/*
 * Given a string, find the length of the longest substring without repeating characters.

  Examples:

  Given "abcabcbb", the answer is "abc", which the length is 3.

  Given "bbbbb", the answer is "b", with the length of 1.

  Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.

  Subscribe to see which companies asked this question

 */
public class Three {
  public int lengthOfLongestSubstring(String s) {
    byte[] chars=s.getBytes();
    HashMap<Byte,Integer> map=new HashMap<>();
    int i=0,j=i;
    int max=0;
    while (j<chars.length){
      if (!map.containsKey(chars[j])){
        map.put(chars[j], j);
      } else {
        if (map.size()>max){
          max=map.size();
        }
        int newI=map.get(chars[j])+1;
        for (int ii=i;ii<newI;ii++){
          map.remove(chars[ii]);
        }
        i=newI;
        map.put(chars[j], j);
      }
      j++;
    }
    if (map.size()>max){
      max=map.size();
    }
    return max;//TODO
  }
}

File: javascript/three.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
	if (!s || s.length===0){
		return 0;
	}

	var map={};
	var len=0;
	var maxLen=1;
    var start=0;

    for (var i=start; i<s.length; i++){
    	c=s[i];
    	if (map[c]!==undefined && map[c]>=start){
    		start=map[c]+1;
    	}
		len=i-start+1;

    	if (len>maxLen){
    		maxLen=len;
    	}

    	map[c]=i;
    }
    return maxLen;
};

console.log(lengthOfLongestSubstring("abcabcbb"));
console.log(lengthOfLongestSubstring("bbbbb"));
console.log(lengthOfLongestSubstring("pwwkew"));
console.log(lengthOfLongestSubstring("c"));
console.log(lengthOfLongestSubstring("au"));